### PR TITLE
Fix Namespace loading in PyYAML 5.4.x

### LINKED ISSUE #6666: Remove requirement of PyYAML!=5.4.x
## 🚀 Feature
Remove dependency requirement of PyYAML!=5.4.x

### Motivation

According to [safety](https://pypi.org/project/safety/) PyYAML versions below 5.4 have a security vulnerability which means that our pre-commit hooks don't allow upgrading to a newer versions of lightning.

```
-> pyyaml, installed 5.3.1, affected <5.4, id 39611
A vulnerability was discovered in the PyYAML library in versions before 5.4, where it is susceptible to 
arbitrary code execution when it processes untrusted YAML files through the full_load method or with 
the FullLoader loader. Applications that use the library to process untrusted input may be vulnerable 
to this flaw. This flaw allows an attacker to execute arbitrary code on the system by abusing the 
python/object/new constructor. This flaw is due to an incomplete fix for CVE-2020-1747. See CVE-2020-14343.
```

### Pitch

Figure out why 5.4.x doesn't work for lightning and remove this requirement if possible.

### LINKED ISSUE #7291: Allow PyYAML 5.4
## 🚀 Feature

Right now the requirement file blocks the installation of PyYAML >=5.4. Would it be possible to remove this constrain?

### Motivation

There is CVE out for PyYAML 5.3.1 [link]( https://nvd.nist.gov/vuln/detail/CVE-2020-1747) and this makes this version unusable in regulated environments where we are required by law to update to patched versions. Now combined with the recent PIP changes about how versions are handled it means that we cannot use pytorch lighining as we required PyYAML >5.4.

### Pitch

Allow PyTorch-Lightning to be installed with PyYAML > 5.4

### PR DESCRIPTION
**Pull request by chris-boson**
## What does this PR do?

This fixes Namespace loading with PyYAML 5.4.x, which was restricted due to a security vulnerability.

Related issues:
https://github.com/PyTorchLightning/pytorch-lightning/pull/5606
https://github.com/PyTorchLightning/pytorch-lightning/pull/5619

Fixes #6666
fixes #7291

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/saving.py [hunk 1]
```diff
@@ -350,7 +350,7 @@ def load_hparams_from_yaml(config_yaml: str, use_omegaconf: bool = True) -> Dict
         return {}
 
     with fs.open(config_yaml, "r") as fp:
-        hparams = yaml.load(fp, Loader=yaml.UnsafeLoader)
+        hparams = yaml.load(fp, Loader=yaml.Loader)
```
Comment (kaushikb11) [inline pytorch_lightning/core/saving.py:353]: @chris-boson Why replace this?

**Comment by Borda** [comment]
@chris-boson pls target master as default branch

File: pytorch_lightning/core/saving.py (hunk 1 above)
Comment (chris-boson) [reply pytorch_lightning/core/saving.py:353]: Actually I think this does just default to UnsafeLoader, but it doesn't seem to cause problems with pyyaml 5.4.x functionality wise.
I think if we want to use SafeLoader we need to add our own Loader and add constructors https://github.com/yaml/pyyaml/issues/482#issuecomment-765737796

Comment (chris-boson) [reply pytorch_lightning/core/saving.py:353]: Similar to this https://stackoverflow.com/a/52586430

**Comment by chris-boson** [comment]
@Borda Are we ok with just removing `PyYAML!=5.4.x` from requirements and continue using `UnsafeLoading`?  What were the concerns with excluding this version? I think in terms of safety that would be equivalent  to using `PyYAML<5.4`.

Alternatively we need to add our own `SafeLoader` and constructors https://github.com/PyTorchLightning/pytorch-lightning/pull/6673#discussion_r601372959.

File: requirements.txt [hunk 2]
```diff
@@ -4,7 +4,7 @@ numpy>=1.16.6
 torch>=1.4
 future>=0.17.1  # required for builtins in setup.py
 # pyyaml>=3.13
-PyYAML>=5.1, !=5.4.*  # OmegaConf requirement >=5.1
+PyYAML>=5.1  # OmegaConf requirement >=5.1
```
Comment (omry) [inline requirements.txt:7]: OmegaConf declares it's version dependency on pyyaml so this line should no longer been needed (At some point in the far past it didn't, which was the reason for adding this in the first place).

https://github.com/omry/omegaconf/blob/master/requirements/base.txt

File: pytorch_lightning/core/saving.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/core/saving.py:353]: @Borda

**Comment by Borda** [comment]
> I think in terms of safety that would be equivalent to using `PyYAML<5.4`

we do not want any future versions like 6.x or so... so rather ban specific version if needed

File: pytorch_lightning/core/saving.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/core/saving.py:353]: we are fine with declare own Loader, @chris-boson mind send a suggestion?

**Comment by Borda** [comment]
@chris-boson how is it going here?

**Comment by carmocca** [comment]
So to use SafeLoader we would need to add constructors for every object serialized?

Should we just use the UnsafeLoader then? AFAIK this was the case before, and we can't guarantee what will be saved to `hparams`

**Comment by chris-boson** [comment]
> So to use SafeLoader we would need to add constructors for every object serialized?
> 
> Should we just use the UnsafeLoader then? AFAIK this was the case before, and we can't guarantee what will be saved to `hparams`

Yeah, at least it wouldn't be any worse than it was before. So effectively we'd just take out the PyYAML!=5.4.* from requirements

**Comment by carmocca** [comment]
Exactly, @Borda, do you agree?

**Comment by Borda** [comment]
> So effectively we'd just take out the PyYAML!=5.4.* from requirements

but isn't that change in PyYAML for the future, is there an expectation that 5.5 will be fine?

**Comment by chris-boson** [comment]
> > So effectively we'd just take out the PyYAML!=5.4.* from requirements
> 
> but isn't that change in PyYAML for the future, is there an expectation that 5.5 will be fine?

We can put <=5.4.1?

**Review by awaelchli — APPROVED** [review]

**Comment by carmocca** [comment]
@chris-boson what's blocking this?

**Comment by Borda** [comment]
> @chris-boson what's blocking this?

seems like just failing tests...

**Review by Borda — APPROVED** [review]
can we update our YAML usage to support also newer versions?

**Review by ananthsub — APPROVED** [review]

**Pull request merged**